src/libtsduck/dtv/analysis/tsTr101Analyzer.cpp

```diff
@@ -0,0 +1,503 @@
+        service->pat_err.count++;
+        service->pat_err2.count++;
+    }
+
+    if (service->_type == ServiceContext::Pmt && table.tableId() != TID_PMT) {
+        service->pmt_err.count++;
+        service->pmt_err2.count++;
+    }
+
+    if (table.tableId() == TID_PAT && service->_type == ServiceContext::Pat) {
+        service->_type = ServiceContext::Pat;
+        if (service->_last_table_ts != INVALID_PCR) {
+            auto diff = long(_currentTimestamp - service->_last_table_ts);
+            service->pat_err.pushSysClockFreq(diff);
+            service->pat_err2.pushSysClockFreq(diff);
+            if (diff > PAT_INTERVAL) {
+                service->pat_err.count++;
+                service->pat_err2.count++;
+            }
+        }
+        service->_last_table_ts = _currentTimestamp;
+    }
+    else if (table.tableId() == TID_PMT && service->_type == ServiceContext::Pmt) {
+        if (service->_last_table_ts != INVALID_PCR) {
+            auto diff = long(_currentTimestamp - service->_last_table_ts);
+            service->pmt_err.pushSysClockFreq(diff);
+            service->pmt_err2.pushSysClockFreq(diff);
+            if (diff > PMT_INTERVAL) {
+                service->pmt_err.count++;
+                service->pmt_err2.count++;
+            }
+        }
+        service->_last_table_ts = _currentTimestamp;
+    }
+    else if (table.tableId() == TID_CAT) {
+        _lastCatIndex = _packetIndex;
+
+        for (auto & [_, s2] : _services) {
+            s2->cat_error = 0;
+        }
+    }
+}
+void ts::TR101_290Analyzer::handleInvalidSection(SectionDemux& demux, const DemuxedData& data) {
+    auto _service = getService(data.sourcePID());
+    // todo: it is not guaranteed that it is a CRC error that causes this issue.
+    _service->crc_error++;
+}
+
+std::shared_ptr<ts::TR101_290Analyzer::ServiceContext> ts::TR101_290Analyzer::getService(PID pid)
+{
+    auto it = _services.find(pid);
+    if (it == _services.end()) {
+        auto service = std::make_shared<ServiceContext>(ServiceContext {
+            ._pid = pid,
+            ._type = ServiceContext::Unassigned});
+        _services.insert(std::make_pair(pid, service));
+        return service;
+    }
+    return it->second;
+}
+
+void ts::TR101_Options::defineArgs(Args& args)
+{
+    json.defineArgs(args, true, u"JSON");
+
+    args.option(u"show-report", 0, Args::NONE);
+    args.help(u"show-report", u"Show an TR 101-290 analyzer report before exiting. By default this is enabled, unless JSON is set.");
+}
+
+bool ts::TR101_Options::loadArgs(DuckContext& duck, Args& args)
+{
+    if (!json.loadArgs(duck, args)) return false;
+
+    show_report = !json.useJSON() || args.present(u"show-report");
+    return true;
+}
+void ts::TR101_290Analyzer::processPacket(ServiceContext& ctx, const TSPacket& pkt, const TSPacketMetadata& mdata) const
+{
+
+    // todo: do TS packets ever arrive with invalid sync bytes this deep in?
```

Comment:
I tried to test some Sync errors, but they didn't make their way all the way down into the plugin. Is there a way to query for these errors?

Reply:
Not easy or even meaningful to test with a generic framework such as TSDuck. This is something to test at demodulator level. And even at modulator level, it is not really meaningful to differentiate corruption of that byte, compared to other byte. This is a bit artificial.

At tsp level, packets with corrupted sync byte will probably never be output by the receiver or input plugin.

Reply:
So, I think testing for this is still important.

 TR 101 290 defines this as:
>  The indicator "Sync_byte_error" is set as soon as the correct sync byte (0x47) does not appear after 188 or 204 bytes.

In a file this is super easy to produce by just deleting a couple of bytes from the middle of a file recording.

 I don't think this error would ever trigger in practice in an IP stream, but I could see some potential condition, where if you converted IP packets to a byte stream and then back to IP while taking loss in the byte stream, it can be possible for the conversion back to IP to cause a sync loss.

In both cases, TSDuck would be the "demodulator", especially if you're performing more complex operations, such as a `regulate`.

Reply:
@lelegard pining again, as I think my previous comment got missed.